src/course/thread-net/control-app.md

```diff
@@ -0,0 +1,213 @@
+still cause issues. For instance, when a function called from within a
+callback `yield`s internally, it can unexpectedly be run _within_ the
+execution of the function. This can in turn break the function's
+semantics. Thus, it is good practice to restrict callback handler code
+to only non-blocking operations.
+
+As such, we instead architect our controller and sensor application
+interactions using two callbacks and an asynchronous timer. It will
+work as follows:
+
+1. The `main` function will request the `sensor` app to provide a
+   temperature reading, and thus issue an IPC callback.
+2. The IPC client callback will save the temperature value, and
+   request a timer callback in 250 ms.
+3. The timer callback will request an IPC service call from the sensor
+   app, going back to step 2.
+
+As such, this loop does not execute any blocking / `yield`ing
+operations in any callback. It also moves all timing / scheduling
+logic out of the applications main loop, which can instead look like
+this:
+
+```
+int main(void) {
+  // Send initial IPC request
+
+  // Yield in a loop, allowing callbacks to be run:
+  for (;;) {
+    yield();
+  }
+}
+```
+
+The final piece of the puzzle is to run blocking code in response to
+these callbacks, but outside of the callback handlers themselves. For
+this, Tock provides the `yield_for` function: it `yield`s the
+application, _until_ a certain condition is met. For instance, the
+controller application sets the `callback_event` boolean variable to
+`true` every time a callback is run. When we want to wait on this
+event in our main function, we can use the following logic:
+
+```
+// Shared variable to signal whether a callback has fired:
+bool callback_event = false;
+
+void ipc_callback() {
+  // Indicate that a callback has fired:
+  callback_event = true;
+}
+
+int main(void) {
+  // Send initial IPC request
+
+  // Yield in a loop, allowing callbacks to be run:
+  for (;;) {
+    // Wait for callback_event to be true:
+    yield_for(&callback_event);
+	// Reset callback_event to false for the next iteration:
+	callback_event = false;
+
+	// This code is executed whenever one or more callbacks have
+	// fired. It can be long running and yield and will not be
+	// re-entered:
+	// ...
+  }
+}
+```
+
+> **EXERCISE:** The `03_controller_screen` checkpoint already contains
+> the logic outlined above. Extend the main function to, in response
+> to a callback, write the current temperature on a screen. You might
+> find it useful to split this code out into a different function.
+
+Finally, we will wire up this application to the OpenThread network to
+send the current temperature setpoint to all other control units, and
+retrieve an average value back. We provide some useful scaffolding for
+this in the next checkpoint, so it is advisable to either switch to
+that, or copy the commented out function signatures for OpenThread
+communication and integration at this point:
```

Comment:
I'm a little confused by what we mean here. My understanding of our structure is:

1. sensor app
2. controller / sensor ipc / screen display
3. openthread
4. add logic to controller for openthread ipc

Are we planning to have another submodule after OpenThread or are we just going to provide this controller logic?